### ISSUE TITLE
[Bug]: CHGNet predictions are not supercell-invariant for periodic self-image bonds within the three-body cutoff

### ISSUE THREAD

**Issue by wakamiya0315** [materialyzeai/matgl#839]
**Version**

main @ d930a16c89ce5082ba92aa0bf5076a2e3739d744 (also reproducible in v4.0.3)

**Which OS(es) are you using?**

MacOS, Linux

**What happened?**

In the PyG CHGNet implementation, `create_directed_line_graph()` handles bonds from an atom to its own periodic image (`src == dst` with a nonzero PBC offset) differently from other bonds. The bond angles are correct, but for these neighbors the line-graph convolution reads the features of the opposite directed bond. In an equivalent supercell, the same neighbor has a different atom index and is handled like any other bond. As a result, a unit cell and its supercell can give different energies, forces and stresses for the same crystal.

[Relevant code at d930a16](https://github.com/materialyzeai/matgl/blob/d930a16c89ce5082ba92aa0bf5076a2e3739d744/src/matgl/graph/_compute.py#L469-L512)

The same logic is in the v4.0.3 release and in the earlier DGL implementation ([v1.3.0](https://github.com/materialyzeai/matgl/blob/a6ebaebd6a0021e14d49d001103508c4980d94da/src/matgl/graph/compute.py#L258-L321), [v3.0.5](https://github.com/materialyzeai/matgl/blob/fcf19dc02ac6019a30c538129019ec0adc3abd3d/src/matgl/graph/_compute_dgl.py#L257-L320)).

Consider a center atom c with a periodic image c+o inside `threebody_cutoff`. `CHGNetLineGraphConv.node_update_` reads the src-bond features as `bonds_i = node_features[lg_src]`. For the neighbor at c+o:

| | vector used for the angle | features read (`bonds_i`) |
|---|---|---|
| unit cell (self-image bond, sign +1) | `+bond_vec(c→c+o)` | bond c→c+o (center → neighbor) |
| supercell (c+o is another atom c′, sign −1) | `−bond_vec(c′→c)` | bond c′→c (neighbor → center), i.e. c→c−o by translation |

The angles agree, but the features belong to c→c+o in the unit cell and to c→c−o in the supercell. The initial bond features depend only on distance, so the first block agrees. Once the bond features have been updated, h(c→c+o) and h(c→c−o) generally differ unless a site-symmetry operation maps o to −o, and the subsequent layers become cell-representation dependent.

This can occur whenever a nonzero lattice translation places an atom's periodic image within `threebody_cutoff` (3 Å). Examples are displaced, relaxing or MD configurations of small bcc, fcc and hcp metal cells or of the diamond primitive cell. Highly symmetric structures may show no numerical difference. This is separate from [#834](https://github.com/materialyzeai/matgl/issues/834): for the Fe cell below, autograd forces matched central finite differences with both implementations and both checkpoints (maximum error ≤ 7.3e-8 eV/Å, h = 1e-4 Å, float64). The energy itself depends on the chosen cell.

This situation is common in MatPES. In the official MatPES 2025.2 splits, about 20% of the structures have a periodic self-image within 3 Å, that is, a nonzero lattice translation of at most 3 Å, so that every atom in the cell has such bonds:

| Functional | train | valid | test |
|---|---:|---:|---:|
| PBE | 78,125 / 389,870 (20.0%) | 4,315 / 21,659 (19.9%) | 4,340 / 21,660 (20.0%) |
| r2SCAN | 74,896 / 347,889 (21.5%) | 4,151 / 19,327 (21.5%) | 4,071 / 19,328 (21.1%) |

Each entry is the number of such structures / all structures in the split. These are mostly small cells (about 80% have at most 4 atoms), so they contain about 10–11% of all atoms. The count is taken from the lattice alone; on 400 PBE test structures it matched the self-image edges in MatGL's own graph.

I compared the current implementation with a local variant of `create_directed_line_graph()`. The variant treats every src bond, including self-image bonds, as an incoming bond (sign −1), and excludes only the reverse bond, identified by atom indices and the PBC offset. The diagonal (i == j) is kept, because for a self-image bond it pairs the neighbors at −o and +o (a valid 180° triplet). In the current code, this amounts to `include_mask = incoming_to_ca & ~is_backtrack`, without the `shared_src` term, the `torch.where(self_edges_j, ...)` override and `fill_diagonal_(False)`, and a sign of −1 for every src bond. The variant is defined in the snippet below.

Only this function was changed. The pretrained weights and all other model code were unchanged; no retraining was performed.

Each 2-atom cell was compared with its 2x2x2 supercell, using the released 1M checkpoints on CPU in float64. Environment: macOS arm64, Python 3.13, PyTorch 2.14.0, PyG 2.8.0.post1, ASE 3.29.0.

| Model | Structure | Line graph | ΔE (eV/atom) | ΔF (eV/Å) | Δσ (eV/Å³) |
|---|---|---|---:|---:|---:|
| CHGNet-PES-MatPES-PBE-1M-2026.9 | Fe bcc, strained + displaced | Current | 4.48e-05 | 3.95e-03 | 5.72e-05 |
| CHGNet-PES-MatPES-PBE-1M-2026.9 | Fe bcc, strained + displaced | Variant | 0 | 2.25e-14 | 2.58e-16 |
| CHGNet-PES-MatPES-PBE-1M-2026.9 | C diamond, displaced | Current | 8.35e-04 | 6.84e-02 | 2.31e-03 |
| CHGNet-PES-MatPES-PBE-1M-2026.9 | C diamond, displaced | Variant | 0 | 1.05e-14 | 4.93e-16 |
| CHGNet-PES-MatPES-r2SCAN-1M-2026.9 | Fe bcc, strained + displaced | Current | 1.31e-04 | 8.86e-03 | 1.60e-04 |
| CHGNet-PES-MatPES-r2SCAN-1M-2026.9 | Fe bcc, strained + displaced | Variant | 0 | 1.22e-14 | 4.66e-16 |
| CHGNet-PES-MatPES-r2SCAN-1M-2026.9 | C diamond, displaced | Current | 8.61e-04 | 5.99e-02 | 4.05e-03 |
| CHGNet-PES-MatPES-r2SCAN-1M-2026.9 | C diamond, displaced | Variant | 0 | 1.11e-14 | 9.20e-16 |

ΔE is the difference in energy per atom. ΔF is the maximum absolute difference over all Cartesian force components of corresponding atoms, and Δσ the maximum over the six Voigt stress components. float32 gives essentially the same differences for the current implementation (checked for both checkpoints).

Further observations:

- Ideal bcc Fe, diamond C and hcp Co, and a strained 1-atom fcc Cu cell, show no difference (ΔF ≤ 2e-14 eV/Å) with either implementation. A displaced hcp Co cell shows ΔF = 1.8e-03 eV/Å (PBE) with the current implementation and 7e-15 eV/Å with the variant.
- A displaced Si cell with no periodic self-image within 3 Å, and every 2x2x2 and 3x3x3 supercell tested, give bit-identical energy, forces and stress with both implementations. With the variant, the unit-cell result equals the current implementation's supercell result (ΔE = 0, ΔF ≤ 6e-14 eV/Å).
- The angle geometry is the same in both representations: the triplet counts agree (Fe: 364 in the unit cell, 2912 = 8 × 364 in the supercell), and so does the multiset of angles. What differs is which directed bond's features are paired with each angle. Atom features agree through `gc_2` and differ from `gc_3` on.
- The 2025.2.10 PyG checkpoints behave the same way. For `fe_perturbed` in `tests/models/test_chgnet.py`, the pinned reference energy matches the unit-cell prediction, which differs from the prediction for its 2x2x2 supercell by 1.2e-4 eV/atom (PBE) and 2.1e-4 eV/atom (r2SCAN).

In my evaluation on the complete MatPES 2025.2 test files (Linux, NVIDIA H100 MIG, CUDA float32 with TF32 disabled, PyTorch 2.13.0), the variant leaves every structure without a self-image unchanged up to float32 round-off. On the structures with a self-image, it changes the largest force component by a median of 0.016 eV/Å per structure and makes the MAE against DFT slightly higher (force MAE: PBE 100.7 → 103.1 meV/Å, r2SCAN 141.5 → 143.6 meV/Å; whole test set: 85.8 → 86.0 and 121.1 → 121.3 meV/Å). This is consistent with the checkpoints having been trained with the current convention, so keeping the accuracy may require retraining with a consistent convention.

Given these results, would it be appropriate to handle bonds to an atom's own periodic image with the same incoming-bond convention as the other bonds?

**Code snippet**

```python
# Unit cell vs 2x2x2 supercell, with the current line graph and with a local variant
# that handles bonds to an atom's own periodic image like all other bonds.
from unittest.mock import patch

import matgl
import matgl.models._chgnet as chgnet_impl
import numpy as np
import torch
from ase.build import bulk
from matgl.ext.ase import PESCalculator

matgl.float_th = torch.float64

# Replace PBE with r2SCAN (revision 061701e6d946d54aa296442cf087479aa80ff358) to check the other checkpoint.
potential = matgl.load_model(
    "BowenD-UCB/CHGNet-PES-MatPES-PBE-1M-2026.9", revision="b12d29dc48fd03979a688f3701b3225a1bb777fb"
).double().eval()


def create_directed_line_graph_incoming(edge_index, pbc_offset, bond_vec, bond_dist, threebody_cutoff):
    """Local variant: every src bond, including bonds to a periodic self-image, is incoming (sign -1)."""
    with torch.no_grad():
        edge_ids = (bond_dist <= threebody_cutoff).nonzero(as_tuple=False).squeeze(1)
        v_src, v_dst, v_img = edge_index[0][edge_ids], edge_index[1][edge_ids], pbc_offset[edge_ids]
        incoming = v_dst[:, None] == v_src[None, :]
        # exclude only the reverse bond (same atom pair, opposite image)
        backtrack = incoming & (v_src[:, None] == v_dst[None, :]) & torch.all(-v_img[:, None] == v_img[None, :], dim=2)
        # the diagonal is kept: for a self-image bond it pairs the neighbors at -o and +o (180 degrees)
        lg_src, lg_dst = (incoming & ~backtrack).nonzero(as_tuple=True)
        sign = bond_vec.new_full((edge_ids.numel(), 1), -1.0)
    return torch.stack([lg_src, lg_dst]), bond_vec[edge_ids], bond_dist[edge_ids], pbc_offset[edge_ids], sign


fe = bulk("Fe", "bcc", a=2.87, cubic=True)  # 2 atoms; |a| = 2.87 A < 3 A three-body cutoff
fe.set_cell(fe.cell @ [[1.012, 0.008, -0.003], [0.008, 0.991, 0.006], [-0.003, 0.006, 1.005]], scale_atoms=True)
fe.positions[0] += [0.035, -0.021, 0.017]
c = bulk("C", "diamond", a=3.567)  # 2-atom primitive cell; |a| = 2.52 A
c.positions[0] += [0.03, -0.02, 0.01]


def compare(unit):
    unit = unit.copy()
    sc = unit.repeat((2, 2, 2))  # the first len(unit) atoms of sc are the original atoms
    unit.calc = PESCalculator(potential, stress_unit="eV/A3", use_voigt=True)
    sc.calc = PESCalculator(potential, stress_unit="eV/A3", use_voigt=True)
    de = abs(unit.get_potential_energy() / len(unit) - sc.get_potential_energy() / len(sc))
    df = np.abs(unit.get_forces() - sc.get_forces()[: len(unit)]).max()
    ds = np.abs(unit.get_stress() - sc.get_stress()).max()
    return f"|dE/atom| = {de:.2e} eV, max|dF| = {df:.2e} eV/A, max|dstress| = {ds:.2e} eV/A^3"


for name, atoms in (("Fe bcc, strained + displaced", fe), ("C diamond, displaced", c)):
    print(f"Current MatGL, {name}: {compare(atoms)}")
    with patch.object(chgnet_impl, "create_directed_line_graph", create_directed_line_graph_incoming):
        print(f"Local variant, {name}: {compare(atoms)}")
```

**Log output**

No exception was raised. Output of the snippet (PBE):

```text
Current MatGL, Fe bcc, strained + displaced: |dE/atom| = 4.48e-05 eV, max|dF| = 3.95e-03 eV/A, max|dstress| = 5.72e-05 eV/A^3
Local variant, Fe bcc, strained + displaced: |dE/atom| = 0.00e+00 eV, max|dF| = 2.25e-14 eV/A, max|dstress| = 2.58e-16 eV/A^3
Current MatGL, C diamond, displaced: |dE/atom| = 8.35e-04 eV, max|dF| = 6.84e-02 eV/A, max|dstress| = 2.31e-03 eV/A^3
Local variant, C diamond, displaced: |dE/atom| = 0.00e+00 eV, max|dF| = 1.05e-14 eV/A, max|dstress| = 4.93e-16 eV/A^3
```

**Comment by shyuep** [comment]
@bowen-bd For your review pls. Thanks.
